packages/dx-scheduler-core/src/plugins/common/computeds.ts

```diff
@@ -22,54 +30,57 @@ export const dayScale = (
   return result;
 };
 
-export const timeScale = (
+export const timeScale: TimeScaleFn = (
   currentDate,
   firstDayOfWeek,
   startDayHour,
   endDayHour,
   cellDuration,
   excludedDays,
 ) => {
-  const result = [];
-  const startViewDate = firstDayOfWeek !== undefined
+  const result: TimeScale[] = [];
+  const startDateOfView = firstDayOfWeek !== undefined
     ? calculateFirstDateOfWeek(currentDate, firstDayOfWeek, excludedDays)
     : currentDate;
-  const left = moment(startViewDate).startOf('hour').hour(startDayHour);
-  const right = moment(startViewDate).startOf('hour').hour(endDayHour);
+  const left = moment(startDateOfView as Date).startOf('hour').hour(startDayHour);
+  const right = moment(startDateOfView as Date).startOf('hour').hour(endDayHour);
 
   while (left.isBefore(right)) {
     const startDate = left.toDate();
     left.add(cellDuration, 'minutes');
     result.push({ start: startDate, end: left.toDate() });
   }
-  result[result.length - 1].end = subtractSecond(result[result.length - 1].end);
+  result[result.length - 1].end = subtractSecond(result[result.length - 1].end) as Date;
   return result;
 };
 
-export const availableViewNames = (viewNames, viewName) => {
+export const availableViewNames: PureComputed<
+  [ViewName[], ViewName], ViewName[]
+> = (viewNames, viewName) => {
   if (!viewNames) return [viewName];
   if (viewNames.findIndex(view => viewName === view) === -1) {
     const nextViewNames = viewNames.slice();
     nextViewNames.push(viewName);
     return nextViewNames;
-  } return viewNames;
+  }
+  return viewNames;
 };
 
-export const viewCellsData = (
+export const viewCellsData: ViewCellsDataFn = (
   currentDate, firstDayOfWeek,
   dayCount, excludedDays,
-  startDayHour, endDayHour, cellDuration,
-  currTime,
+  startDayHour, endDayHour,
+  cellDuration, currTime,
 ) => {
   const days = dayScale(currentDate, firstDayOfWeek, dayCount, excludedDays);
   const times = timeScale(
     currentDate, firstDayOfWeek, startDayHour, endDayHour, cellDuration, excludedDays,
   );
-  const currentTime = moment(currTime);
+  const currentTime = moment(currTime as CurrentTime);
 
-  const cells = [];
+  const cells: ViewCell[][] = [];
   times.forEach((time) => {
```

Comment:
Hey, why don't we use `array.reduce` here? And here `days.forEach((day) => { }` 🤔?

Reply:
Why `reduce` better then `ferEach`?

Reply:
We prefer [functional programming](https://medium.com/jsguru/javascript-functional-programming-map-filter-and-reduce-846ff9ba492d). ☝️

Reply:
Oh... 😢